### PR TITLE
analyzer: add PortNameAnalyzer

### PR DESCRIPTION
**Pull request by yxue**
Fix https://github.com/istio/istio/issues/19387

### CODE REVIEW

**Comment by yxue** [comment]
@esnible I added a document in the UX WG and a github issue. 

Document: https://docs.google.com/document/d/1HkioME7IyoQULbJXRVoMha3f-JU9IJtrFrukm-kgECo/edit#
Issue: https://github.com/istio/istio/issues/19387

**Comment by yxue** [comment]
@esnible as discussed in the issue. I added the check to the port name analyzer.

File: galley/pkg/config/analysis/analyzers/testdata/service-no-port-name.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# No port name
```
Comment (sushicw) [inline galley/pkg/config/analysis/analyzers/testdata/service-no-port-name.yaml]: 1. Please annotate the test cases here with comments indicating what should create messages (or not) and why.
2. Please also add a negative test case that shouldn't generate any messages.

Comment (yxue) [reply galley/pkg/config/analysis/analyzers/testdata/service-no-port-name.yaml]: Done

**Comment by yxue** [comment]
@esnible could you help to review the PR? Thanks!

**Comment by yxue** [comment]
For release 1.3.x and 1.4, port name is not required for outbound. For 1.5, port name is not required for both inbound and outbound. And sniffing for outbound doesn't have any security risk.

**Comment by yxue** [comment]
@istio/wg-networking-maintainers can you please review the PR? Thanks!

File: galley/pkg/config/analysis/analyzers/testdata/service-port-name.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# If port is unnamed or port name doesn't follow <protocol>[-<suffix>], the analyzer will report warning.
+apiVersion: v1
+kind: Service
+metadata:
+name: my-service
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply galley/pkg/config/analysis/analyzers/testdata/service-port-name.yaml]: given this is invalid but the tests pass, is this testing the right thing?

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 3]
```diff
@@ -209,3 +209,16 @@ messages:
     args:
       - name: deployment
         type: string
+
+  - name: "PortNameIsNotUnderNamingConvention"
+    code: IST0118
+    level: Warning
+    description: "Port name is not under naming convention. Protocol detection is applied to the port."
```
Comment (howardjohn) [inline galley/pkg/config/analysis/msg/messages.yaml]: Why is this a warning? The whole point of protocol detection was that a user does not need to care about the port name, but now we are going to spit out 1000s of messages about the port name whne they run this?

File: galley/pkg/config/analysis/analyzers/testdata/service-port-name.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (yxue) [reply galley/pkg/config/analysis/analyzers/testdata/service-port-name.yaml]: Yes, this follows convention and the test checks there is no warning message. I think the check is valid even though I don't know the effect of unformatted yaml file.

File: galley/pkg/config/analysis/msg/messages.yaml (hunk 3 above)
Comment (yxue) [reply galley/pkg/config/analysis/msg/messages.yaml]: I think make the cmd optional and set the level Warning would be good. If we cannot make the cmd as optional, I will change this to Info level.

**Comment by yxue** [comment]
> Can we make this opt in or someting? This is not relevant for most users. Only those on the "hardening istio cluster" path

@sushicw can we make the command optional?

File: galley/pkg/config/analysis/analyzers/testdata/service-port-name.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (howardjohn) [reply galley/pkg/config/analysis/analyzers/testdata/service-port-name.yaml]: Oh the comment made me think this was supposed to not... 

The formatting is not just cosmetic

[the remark this answers is not in the document]
Comment (yxue) [reply galley/pkg/config/analysis/analyzers/testdata/service-port-name.yaml]: I guess the analyzer won't apply or parse the yaml correctly so there is no warning message. I didn't dig the problem here.

File: galley/pkg/config/analysis/msg/messages.yaml (hunk 3 above)
Comment (yxue) [reply galley/pkg/config/analysis/msg/messages.yaml]: There is an open issue to make the cmd optional https://github.com/istio/istio/issues/17443. The milestone is 1.5. I think we could keep this message level warning.

Comment (yxue) [reply galley/pkg/config/analysis/msg/messages.yaml]: Or we can turn this to warning when  #17443 is done.

**Comment by sushicw** [comment]
> > Can we make this opt in or someting? This is not relevant for most users. Only those on the "hardening istio cluster" path
> 
> @sushicw can we make the command optional?

Summarizing offline discussion:

Not yet.

https://github.com/istio/istio/issues/17443 will allow specifying which analyzers to run, but does not currently have in-scope the idea of "non-default" analyzers that don't run unless you specifically ask for them. Perhaps this is a use case where having some analyzers disabled by default would make sense?

WARN makes the most sense semantically (encouraging the more secure practice is good, right?) but I think we'd only want to have this pop up to users if there was an easy & straightforward way for them to acknowledge & suppress the message. Expecting them to go through and add port names everywhere to make the message go away seems a bit much.

Making it INFO level would mean it still shows up in `istioctl analyze` output by default, but wouldn't return a nonzero exit code. That's only slightly better, really. 

https://github.com/istio/istio/issues/17703 tracks adding the ability to suppress messages, which is probably closer to what we need for this. If the user could (running one command to add an annotation) suppress a particular code, would that be simple enough for them?


So, to summarize the summary, strategy options on the table:
* Have this analyzer not run by default, but be available for explicit opt-in.
* Have this analyzer run by default, but the message is easy to suppress if you don't care.


@selmanj I bet you have some thoughts here.

**Comment by sushicw** [comment]
P.S. I guess one important question to make sure we're aligned on here is how hard we want to push explicit port names as a "best practice."

**Comment by selmanj** [comment]
It's extremely confusing to me that Istio has protocol detection which means I don't have to label my ports, but then I see a message saying I should label my ports as a best practice. The message also doesn't explain why labeling is required (it's a potential security issue).

Some users won't care about the security issue, and just want things to work (maybe they're not using authn policies at all). Those users are benefitting from using protocol detection because it's less work on their part. They will be annoyed by this message.

Some users will care about the security issue. For users who care, why would they continue using protocol detection if they have to label all their ports anyway?

**Comment by selmanj** [comment]
(Also, I agree that this analyzer heavily motivates the use case of allowing users control to turn analyzers on/off. Focusing on that work today since this PR forces the issue).

**Comment by yxue** [comment]
> It's extremely confusing to me that Istio has protocol detection which means I don't have to label my ports, but then I see a message saying I should label my ports as a best practice. The message also doesn't explain why labeling is required (it's a potential security issue).

My assumption is that when user runs this analyzer cmd, they want to transit from unnamed ports to named ports to harden their cluster and they should understand the message and potential security issues caused by unnamed ports. 

> Some users will care about the security issue. For users who care, why would they continue using protocol detection if they have to label all their ports anyway?

Users might want to harden their clusters from the state when unnamed ports are used.

**Comment by howardjohn** [comment]
> My assumption is that when user runs this analyzer cmd, they want to transit from unnamed ports to named ports to harden their cluster and they should understand the message and potential security issues caused by unnamed ports.



How doe sthis related to the other istioctl command to show unnamed ports? Does this replace it?

File: galley/pkg/config/analysis/msg/messages.yaml [hunk 4]
```diff
@@ -209,3 +209,16 @@ messages:
     args:
       - name: deployment
         type: string
+
+  - name: "PortNameIsNotUnderNamingConvention"
+    code: IST0118
+    level: Info
```
Comment (howardjohn) [inline galley/pkg/config/analysis/msg/messages.yaml]: Is info shown to users by default?

**Comment by yxue** [comment]
> How doe sthis related to the other istioctl command to show unnamed ports? Does this replace it?

It's not clear to me. It's not related to other istioctl command to show unnamed ports

File: galley/pkg/config/analysis/msg/messages.yaml (hunk 4 above)
Comment (yxue) [reply galley/pkg/config/analysis/msg/messages.yaml]: I am not familiar with the istioctl analyzer. I think info is also shown to users by default but maybe not highlighted.

**Comment by howardjohn** [comment]
Doesn't it do the same thing?

**Comment by yxue** [comment]
Do we have istioctl cmd to show unnamed ports?

**Comment by howardjohn** [comment]
oh I thought this: https://github.com/istio/istio/issues/19387 was also implemented. Makes sense.

**Comment by howardjohn** [comment]
As I user, when I get an error like:

```
Info [IST0118] (Service httpbin.default) Port name foo-http (port: 80, targetPort: 80) doesn't follow the naming convention of Istio port.
```

I have no clue what the impact of this or what I can do to resolve this.

Its also not clear if this matters. Should we only run this check if protocol detection isdisabled?

**Comment by yxue** [comment]
> As I user, when I get an error like:
> 
> ```
> Info [IST0118] (Service httpbin.default) Port name foo-http (port: 80, targetPort: 80) doesn't follow the naming convention of Istio port.
> ```
> 
> I have no clue what the impact of this or what I can do to resolve this.
> 
> Its also not clear if this matters. Should we only run this check if protocol detection isdisabled?

I think users could run the analyzer tool at any time, either of these two transitions could happen:

turn on protocol detection -> run analyzer tool -> make the change -> turn off protocol detection
turn on protocol detection -> turn off protocol detection -> run analyzer tool -> make the change 

I don't think we should limit the check if protocol detection is disabled or not. 

And if istioctl analyzer support optional command, this problem would go away. As it's expected to be support in 1.5, we can leave the PR as it is.

File: galley/pkg/config/analysis/msg/messages.yaml (hunk 4 above)
Comment (sushicw) [reply galley/pkg/config/analysis/msg/messages.yaml]: Info is shown by default, but won't cause the command to fail with a nonzero exit code.

**Comment by sushicw** [comment]
> As I user, when I get an error like:
> 
> ```
> Info [IST0118] (Service httpbin.default) Port name foo-http (port: 80, targetPort: 80) doesn't follow the naming convention of Istio port.
> ```
> 
> I have no clue what the impact of this or what I can do to resolve this.
> 
> Its also not clear if this matters. **Should we only run this check if protocol detection isdisabled?**

As it's been explained to me, even with protocol detection enabled it's better security to name the ports. (Which seems like a glaring deficiency in protocol detection's usefulness, TBH: why are we pushing something less secure?). 

If we can come up with a 1-sentence explanation of how to fix this, it's worth adding to the description. Otherwise, we'll have to rely on the docs for the error code.

**Comment by yxue** [comment]
@howardjohn @esnible ping

**Comment by howardjohn** [comment]
> As it's been explained to me, even with protocol detection enabled it's better security to name the ports. (Which seems like a glaring deficiency in protocol detection's usefulness, TBH: why are we pushing something less secure?).

Some people care more about UX than security. Others do not. If you are interested in never ending debates about these topics let me know and we can include you :slightly_smiling_face:

**Comment by kyessenov** [comment]
The purpose of the tool is rather confusing (we tell people to use sniffing so that they can name ports and not use sniffing???). Please update the link to the design doc since it's empty https://docs.google.com/document/d/1HkioME7IyoQULbJXRVoMha3f-JU9IJtrFrukm-kgECo/edit#.

**Comment by yxue** [comment]
The design doc was for adding a new istioctl command at the very beginning but people recommend to add the cmd to istioctl analyzer directly. By default, people can use sniffing and we provide tools to detect ports that are using sniffing and user can switch those ports to named ports for security.
